docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+
+
+def least_squares_sphere(params: np.ndarray) -> dict[str, Any]:
+    out = {"root_contributions": params, "p_mean": params.mean(), "p_std": params.std()}
+    return out
+```
+
+**Things we want to keep**
+
+- The fact that `params` can be arbitrary pytrees makes estimagic flexible and popular.
+  We do not need to restrict this type in any way because flattening the pytree gives us
+  a very precise type no matter how complicated the tree was.
+- Using the same criterion function for scalar, likelihood and least-squares optimizers
+  makes it easy to try out and compare very different algorithms with minimal code
+  changes.
+- We do not need to restrict the type of additional arguments of the criterion function.
+- The simplest form of our criterion functions is also compatible with scipy.optimize
+
+**Problems**
+
+- Most users of estimagic find it hard to write criterion functions that return the
+  correct dictionary. Therefore, they don't use the logging feature and we often get
+  questions about specifying least-squares problems correctly.
+- Internally we can make almost no assumptions about the output of a criterion function,
+  making the code that processes the criterion output very complex and full of if
+  conditions.
+- The best typing information we could get for the output of the criterion function is
+  `float | dict[str, Any]`, which is not very useful.
+- We only know if the specified criterion function is compatible with the selected
+  optimizer after we evaluate it once. This means that errors for users are raised very
+  late.
+- While optional, in least-squares problems it is possible that a user specifies
+  `root_contributions`, `contributions` and `value` even though any of them could be
+  constructed out of the `root_contributions`. This means we need to check that all of
+  them are compatible.
+- The name `criterion` is not used very widely in the Python optimization ecosystem.
+
+#### Proposal
+
+The output of the criterion function becomes `float | FunctionValue`. There are
+decorators that help the user to write valid criterion functions without making an
+explicit instance of `FunctionValue`.
+
+The first two previous examples remain valid. The third one will be deprecated and
+should be replaced by:
+
+```python
+@em.mark.least_squares
+def least_squares_sphere(params: np.ndarray) -> em.FunctionValue:
+    out = FunctionValue(
+        residuals=params, info={"p_mean": params.mean, "p_std": params.std()}
+    )
+    return out
+```
+
+We can exploit this deprecation cycle to rename `root_contributions` to `residuals`
+which is more in line with the literature. Similarly, `contributions` can be renamed to
+`loglikes`.
+
+Since there is no need to modify instances of `FunctionValue`, it should be immutable.
+
+If a user only wants to express the least-squares structure of the problem without
+logging any additional information, they can only return the least-squares residuals as
```

Comment:
The logging bit is incomprehensible without additional info.

Reply:
Agree, I hope it is clearer now.